searx/plugins/__init__.py

```diff
@@ -77,6 +87,80 @@ def call(self, ordered_plugin_list, plugin_type, request, *args, **kwargs):
         return ret
 
 
+def load_external_plugins(plugin_names):
+    plugins = []
+    for name in plugin_names:
+        logger.debug('loading plugin: {0}'.format(name))
+        try:
+            pkg = import_module(name)
+        except Exception as e:
+            logger.critical('failed to load plugin module {0}: {1}'.format(name, e))
+            exit(3)
+
+        pkg.__base_path = dirname(abspath(pkg.__file__))
+
+        prepare_package_resources(pkg, name)
+
+        plugins.append(pkg)
+        logger.debug('plugin "{0}" loaded'.format(name))
+    return plugins
+
+
+def sync_resource(base_path, resource_path, name, target_dir, plugin_dir):
+    dep_path = join(base_path, resource_path)
+    file_name = basename(dep_path)
+    resource_path = join(target_dir, file_name)
+    if not exists(resource_path) or sha_sum(dep_path) != sha_sum(resource_path):
+        try:
+            copyfile(dep_path, resource_path)
+        except:
+            logger.critical('failed to copy plugin resource {0} for plugin {1}'.format(file_name, name))
+            exit(3)
+
+    # returning with the web path of the resource
+    return join('plugins', plugin_dir, file_name)
+
+
+def prepare_package_resources(pkg, name):
+    plugin_dir = 'plugin_' + name
+    target_dir = join(static_path, 'plugins', plugin_dir)
+    try:
+        makedirs(target_dir, exist_ok=True)
+    except:
+        logger.critical('failed to create resource directory {0} for plugin {1}'.format(target_dir, name))
+        exit(3)
+
+    resources = []
+
+    if hasattr(pkg, 'js_dependencies'):
+        resources.extend(map(basename, pkg.js_dependencies))
+        pkg.js_dependencies = tuple([
+            sync_resource(pkg.__base_path, x, name, target_dir, plugin_dir)
+            for x in pkg.js_dependencies
+        ])
+    if hasattr(pkg, 'css_dependencies'):
+        resources.extend(map(basename, pkg.css_dependencies))
+        pkg.css_dependencies = tuple([
+            sync_resource(pkg.__base_path, x, name, target_dir, plugin_dir)
+            for x in pkg.css_dependencies
+        ])
+
+    for f in listdir(target_dir):
+        if basename(f) not in resources:
+            resource_path = join(target_dir, basename(f))
+            try:
+                remove(resource_path)
+            except:
+                logger.critical('failed to remove unused resource file {0} for plugin {1}'.format(resource_path, name))
+                exit(3)
```

Comment:
FYI: [sys.exit()](https://docs.python.org/3/library/sys.html#sys.exit) exits by raising a [SystemExit](https://docs.python.org/3/library/exceptions.html#SystemExit) .. which can be catched by the outer block :) .. sometimes it make sense to raise such exception (calling sys.exit()). Itis like the `die` command which might know some of us from perl.  In general this exception remains for the application code on the top, libraries (modules) should pass exceptions upwards.